src/main/scala/org/constellation/util/MajorityStateChooser.scala

```diff
@@ -3,55 +3,100 @@ package org.constellation.util
+    val (nodeWithMajority, allSnapshots): (Id, Seq[RecentSnapshot]) = nodeSnapshots.filter {
+      case (id, snaps) => snapshotNodes._2.contains(id)
+    }.maxBy(snaps => snaps._2.length)
+    allSnapshots.filter(_.height <= snapshotNodes._1.height)//todo filter such that, download peers contain current node's
+    // snap/height or rather there is an intersection without missing heights
+  }
 
-  def chooseMajorityState(
-    nodeSnapshots: List[NodeSnapshots],
-    ownHeight: Long,
-    allPeers: Seq[Id]
-  ): OptionT[F, (Seq[RecentSnapshot], Set[Id])] =
-    for {
-      majorState <- chooseMajoritySnapshot(
-        nodeSnapshots.filter(checkIfNodeContainsSnapshotsInConsistentState),
-        ownHeight,
-        allPeers
-      )
-      snapsThoughMaj <- OptionT.fromOption[F](getAllSnapsUntilMaj(majorState._2, nodeSnapshots))
-//      node <- OptionT.fromOption[F](findNode(nodeSnapshots, nodeId))
-      nodeIds <- OptionT.fromOption[F](chooseMajNodeIds(majorState._2, nodeSnapshots))
-
-      _ <- OptionT.liftF(logger.debug(s"Re-download from nodes : ${nodeIds}"))
-    } yield (snapsThoughMaj.sortBy(-_.height), nodeIds.toSet)
-
-  private def chooseMajoritySnapshot(nodeSnapshots: Seq[NodeSnapshots], height: Long,
-                                     allPeers: Seq[Id]) =
-    for {
-      // highestSnapshot <- OptionT.fromOption[F](getHighest(nodeSnapshots))
-      // useHighest <- OptionT.liftF(shouldUseHighest(highestSnapshot, ownHeight))
-      majorSnapshot <- OptionT.fromOption[F](chooseMajorWinner(allPeers, nodeSnapshots))//def chooseMajorWinner(allPeers: Seq[Id])(nodeSnapshots: Seq[NodeSnapshots])
-
-      _ <- OptionT.liftF(
-        logger.debug(
-//          s"The highest snapshot : $highestSnapshot"
-          s"The major snapshot: $majorSnapshot"
-        )
-      )
-    } yield majorSnapshot
+  def chooseMajorWinner(allPeers: Seq[Id], nodeSnapshots: Seq[NodeSnapshots]) = {
+    val snapshotToProposers: Seq[(RecentSnapshot, Seq[(Id, RecentSnapshot)])] = nodeSnapshots.flatMap {
+      case (id, snapshots) => snapshots.map((id, _))
+    }.groupBy {
+      case (_, snapshot) => snapshot
+    }.toList
+    val snapsGroupedByHeight: Map[Long, Seq[(RecentSnapshot, Seq[Id])]] = snapshotToProposers.map {
+      case (recentSnap, idRecentSnapTups) => (recentSnap, idRecentSnapTups.map(_._1))
+    }.groupBy { case (recentSnap, idRecentSnaps) => recentSnap.height }
+    val res = Try {
+      snapsGroupedByHeight.flatMap {
+        case (height, allSnapshotNodes) =>
+          val getMajOpt = selectMajSnap(allPeers.length, allSnapshotNodes)
+          allSnapshotNodes.find(sn => getMajOpt.contains(sn._1)).map(maj => (height, maj))
+      }.maxBy { case (height, maj) => height }
+    }.toOption
+    res
+  }
+}
+
+class MajorityStateChooser[F[_]: Concurrent] {
+  import MajorityStateChooser._
+
+  val logger = Slf4jLogger.getLogger[F]
+  private final val differenceInSnapshotHeightToReDownloadFromLeader = 10
```

Comment:
Shouldn't in `application.conf`?

Reply:
Yep we also need to add API.recentSnapshotNumber too

Reply:
Actually looks like I already did 😂